### PR TITLE
know recipe? always ask group

### LINKED ISSUE #73096: "View recipe" doesn't go to recipes known by ally
### Describe the bug

See the steps to reproduce.

### Attach save file

[TEST hint good-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/15022460/TEST.hint.good-trimmed.tar.gz)

### Steps to reproduce

1. Load attached save
2. open inventory `/`
3. examine pickaxe `e`
4. Try viewing recipe `V`
5. It says you don't know the recipe
6. Open crafting `&`
7. search `f` for `pickaxe` hit `enter`
8. see that the allied NPC next to you knows the recipe
9. Observe that steps 5 and 8 don't sit well together

### Expected behavior

I am not sure about the hint colour (good/neutral). But it should definitely view the recipe.

Colour meaning so far:
- Green: you can craft it
- Yellow: it is possible to craft, but you don't know the recipe
- Red: This item cannot be crafted

We could add another colour, like purple, meaning "Your NPC knows how to craft it, but you don't". I am not 100% sure if this is possible.

I would probably leave it green. You can tell the NPC to craft it if you need to.

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.4291 (22H2)
- Game Version: 0.G-8860-gd51683cb19 [64-bit] master was 73b6cf73b9
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

I can resolve this issue in a PR. Since I know the code, I will probably take the least effort.

I would expect it to be this line:
https://github.com/CleverRaven/Cataclysm-DDA/pull/68118/files#diff-0bb6c7b942a6c15255f935db679c066dfce4d4114fdb99698c1e1a09b7f5a98aR1830

But I don't see anything wrong, so little debugging is needed.

### PR DESCRIPTION
**Pull request by Brambor**
#### Summary
None

#### Purpose of change
fix #73096
~fix . #73123~ Decided against including this.

The colour is green now since somebody from your group can craft it.
image

Additionally:
- Examining an item now gives you what your group can craft from it.
image

You can test this in the #73096 save by smashing right (adjacent bench) and examining the `metal scrap` it drops.

#### Describe the solution

1. Make `Character::get_group_available_recipes` to replace all occurrences of `Characted::get_available_recipes`, since it was always called with `this` inventory and `this` helpers, except for the bugged cases. So this is a unification.
2. Make `Characted::get_available_recipes` protected
	- there are no calls outside Character to it, so we can
	- I cannot think of a case when you would want just your recipes and not recipes of your group (in the public interface), so making `protected` makes the programmers use the public `Character::get_group_available_recipes`

#### Describe alternatives you've considered

> Examining an item now gives you what your group can craft from it.

Change the caption from 'You could use it to craft:' to 'Your group could use it to craft:', should I?

#### Testing

Testing the fix of #73096:
 - I went by the steps, and the crafting window popped up with the pickaxe selected, as expected.

#### Additional context

### CODE REVIEW

**Comment by Hyperseeker** [comment]
I worry that this might get confusing to newer players, or even experienced players who haven't engaged with the NPC crafting system yet. "I can craft this right now" is not the same as "Someone in our little survivor group knows how to craft this when we get home", or even "...right here, right now" if you have a companion on you. The expectation shifts from "If I have to, I can make this" to "I *might* be able to make this in an emergency, but it's not a given".

Also, perhaps clearly delineating the abilities of the player character vs. other survivors of the group paves way towards a system where the player character is not expected to know everything and be capable of everything, which is the current design direction of the game. The earlier the signs start, the earlier the idea settles in, the less confusion ensues when the hard distinction in skills and abilities gets introduced.

Maybe the line in the item menu should be yellow instead of green: "It's possible; see asterisk", rather than "It's possible, all the conditions have been met". Additionally, I believe it would best if there are two lines for craftables: one for "You can craft this without assistance", and one for "Someone in your group can craft this".

**Comment by Brambor** [comment]
To be clear, the group consist only of allies within reach. You can literally open crafting menu `&`, `f`ilter by component (filter = `c:`...), `c`hange crafter and there it is.

If the companion is at home, this nor the crafting menu doesn't consider them.

This (like the crafting GUI) already considered books nearby, this just considers NPCs nearby too (like the crafting GUI).

I don't believe anybody uses this feature anyway (since it was bugged for so long before I fixed it randomly). So there is noone to be confused :D.

**Comment by Hyperseeker** [comment]
> To be clear, the group consist only of allies within reach.

Clearly I was under the wrong impression.

Cool feature!

File: src/iexamine.cpp [hunk 1]
```diff
@@ -7288,8 +7288,7 @@ void iexamine::workbench_internal( Character &you, const tripoint &examp,
                     break;
                 }
                 const recipe &rec = selected_craft->get_making();
-                const inventory &inv = you.crafting_inventory();
-                if( !you.has_recipe( &rec, inv, you.get_crafting_group() ) ) {
+                if( !you.has_recipe( &rec ) ) {
```
Comment (Brambor) [inline src/iexamine.cpp:7291]: Same theory as above. Player is able to continue craft without the skills required.

File: src/iuse.cpp [hunk 2]
```diff
@@ -8347,8 +8347,7 @@ std::optional<int> iuse::craft( Character *p, item *it, const tripoint & )
         return std::nullopt;
     }
     const recipe &rec = it->get_making();
-    const inventory &inv = p->crafting_inventory();
-    if( !p->has_recipe( &rec, inv, p->get_crafting_group() ) ) {
+    if( !p->has_recipe( &rec ) ) {
         p->add_msg_player_or_npc(
```
Comment (Brambor) [inline src/iuse.cpp:8351]: Same as above.

+ Maybe it should differentiate between '"Doesn't know recipe" and "Doesn't have the skills."

**Review by Brambor — COMMENTED** [review]
I probably broke things.

Edit: No, they were always broken.

File: src/crafting.cpp [hunk 3]
```diff
@@ -578,7 +578,7 @@ bool Character::can_make( const recipe *r, int batch_size ) const
 {
     const inventory &crafting_inv = crafting_inventory();
 
-    if( !has_recipe( r, crafting_inv, get_crafting_group() ) ) {
```
[the remark this answers is not in the document]
Comment (Brambor) [reply src/crafting.cpp:581]: Yup, this is true. But you cannot do it just by selecting it in the Crafting GUI.

1. Have nearby all 10 character
2. Set PC skill to all 10
3. craft a recipe
4. Set PC skill to all 0
5. recraft last recipe `-`

File: src/iexamine.cpp (hunk 1 above)
Comment (Brambor) [reply src/iexamine.cpp:7291]: Yup, this is true too.

File: src/iuse.cpp (hunk 2 above)
Comment (Brambor) [reply src/iuse.cpp:8351]: Yup true too

File: src/crafting.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brambor) [reply src/crafting.cpp:581]: You can do this on master too d30757ea5f

File: src/iexamine.cpp (hunk 1 above)
Comment (Brambor) [reply src/iexamine.cpp:7291]: Can do on master d30757ea5f too

File: src/iuse.cpp (hunk 2 above)
Comment (Brambor) [reply src/iuse.cpp:8351]: Can do on master d30757ea5f too...

File: src/crafting.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brambor) [reply src/crafting.cpp:581]: current status on my local branch: not fixed

File: src/iexamine.cpp (hunk 1 above)
Comment (Brambor) [reply src/iexamine.cpp:7291]: current status on my local branch: fixed, but there is no feedback, so needs further work

File: src/iuse.cpp (hunk 2 above)
Comment (Brambor) [reply src/iuse.cpp:8351]: current status on my local branch: **fixed**, but there is no feedback, so **needs further work**

File: src/crafting.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Brambor) [reply src/crafting.cpp:581]: Decided against fixing it here.

File: src/iexamine.cpp (hunk 1 above)
Comment (Brambor) [reply src/iexamine.cpp:7291]: Decided against fixing it here.

File: src/iuse.cpp (hunk 2 above)
Comment (Brambor) [reply src/iuse.cpp:8351]: Decided against fixing it here.

File: tests/crafting_test.cpp [hunk 4]
```diff
@@ -346,7 +346,7 @@ TEST_CASE( "crafting_with_a_companion", "[.]" )
 
         REQUIRE( std::find( helpers.begin(), helpers.end(), &who ) != helpers.end() );
         dummy.invalidate_crafting_inventory();
-        REQUIRE_FALSE( dummy.get_available_recipes( dummy.crafting_inventory(), &helpers ).contains( r ) );
+        REQUIRE_FALSE( dummy.get_group_available_recipes().contains( r ) );
         REQUIRE_FALSE( who.knows_recipe( r ) );
```
[the remark this answers is not in the document]
Comment (Brambor) [reply tests/crafting_test.cpp:350]: Extra checks cannot hurt anything but compile time and read time.

**Pull request merged**